include/revng/Model/Types.h

```diff
@@ -0,0 +1,903 @@
+  "atomic_int_least64_t",
+  "atomic_uint_least64_t",
+  "atomic_int_fast8_t",
+  "atomic_uint_fast8_t",
+  "atomic_int_fast16_t",
+  "atomic_uint_fast16_t",
+  "atomic_int_fast32_t",
+  "atomic_uint_fast32_t",
+  "atomic_int_fast64_t",
+  "atomic_uint_fast64_t",
+  "atomic_intptr_t",
+  "atomic_uintptr_t",
+  "atomic_size_t",
+  "atomic_ptrdiff_t",
+  "atomic_intmax_t",
+  "atomic_uintmax_t",
+  // C Extensions
+  "_Pragma",
+  "asm",
+};
+
+inline constexpr std::array<const char *, 21> PrimitiveTypeNames = {
+  "uint8_t",   "uint16_t",   "uint32_t",  "uint64_t",  "uint128_t", "int8_t",
+  "int16_t",   "int32_t",    "int64_t",   "int128_t",  "bool8_t",   "bool16_t",
+  "bool32_t",  "bool64_t",   "bool128_t", "float16_t", "float32_t", "float64_t",
+  "float80_t", "float128_t", "void",
+};
+
+inline constexpr std::array<const char *, 10> IntegralTypeNames = {
+  "uint8_t", "uint16_t", "uint32_t", "uint64_t", "uint128_t",
+  "int8_t",  "int16_t",  "int32_t",  "int64_t",  "int128_t",
+};
+
+inline bool isPrimitiveReservedName(const std::string &Name) {
+  return llvm::count(PrimitiveTypeNames, Name);
+}
+
+inline bool isIntegralReservedName(const std::string &Name) {
+  return llvm::count(IntegralTypeNames, Name);
+}
+
+inline bool isVoidReservedName(const std::string &Name) {
+  return Name == "void";
+}
+
+inline bool isCReservedKeyword(const std::string &Name) {
+  return llvm::count(CReservedKeywords, Name);
+}
+
+RecursiveCoroutine<std::size_t> size(const flatc::Type *T);
+
+RecursiveCoroutine<std::size_t> size(const flatc::QualifiedType &T);
+
+} // end namespace flatc
+
+/// \brief A primitive type in flatc: sized integers, booleans, floats and void.
+struct flatc::PrimitiveType : public flatc::Type {
+  static constexpr const char *Tag = "!flatc::PrimitiveType";
+
+  PrimitiveType(llvm::StringRef NameRef) : Type(NameRef, TypeKind::Primitive) {}
+  PrimitiveType() : PrimitiveType("") {}
+  ~PrimitiveType() = default;
+
+  PrimitiveType(const PrimitiveType &) = default;
+  PrimitiveType &operator=(const PrimitiveType &) = default;
+
+  PrimitiveType(PrimitiveType &&) = default;
+  PrimitiveType &operator=(PrimitiveType &&) = default;
+
+  static bool classof(const Type *T) {
+    return T->getKind() == TypeKind::Primitive;
+  }
+
+private:
+  // Kind is TypeKind::Primitive. Name is allowed by PrimitiveTypeNames.
+  bool verifyDerived() const debug_function {
+    return getKind() == TypeKind::Primitive and isPrimitiveReservedName(Name);
+  }
+
+  friend struct flatc::Type;
+};
+// Enable introspection for flatc::PrimitiveType
+INTROSPECTION_NS(flatc, PrimitiveType, UUID, Name, Kind);
+
+/// \brief KeyedObjectTraits for std::string based on its value
+template<>
+struct KeyedObjectTraits<std::string>
+  : public IdentityKeyedObjectTraits<std::string> {};
+
+/// \brief An entry in a flatc enum, with a name and a value.
+struct flatc::EnumEntry {
+
+  std::size_t Value;
+  std::string Name;
+  SortedVector<std::string> Aliases;
```

Comment:
What's your reasoning for not just having a vector of `Names` instead of `Name` + `Aliases`?

Reply:
My main concern was about invalidation. If you decide that the first element of `Alisases` is "the name", you could invalidate it by adding a new element to `Aliases` that is sorted before the old one.

Not sure this is actually a problem. But having `Name` as a separate field you have consistency, which is good for debugging, and in general for whatever is user-facing (like the code generated by the decompiler). The name of an entry cannot change unless you change it explicitly.

But I can just go with `Names` if you like it more.

Reply:
I've no strong opinions. It's just a bit more cumbersome to use downstream. For instance, in the UI, we'll have to implement the UI to change both Name and Aliases.

On the other hand, if we don't do it like this, the user cannot specify a "favorite" `Name`, since a `SortedVector` is essentialy a `set` (it's ordered by the string content).

I'll leave it to you.

Reply:
I see you got the point. I thought that leaving to the user the option to set a "favorite" `Name` was good. It's likely a little bit more cumbersome to implement on the UI side, but my feeling is that it has less-unintuitive meaning for the end user. In 95% of the cases I expect the user to want to: "Just rename this entry and be done with it". So the aliases are there mostly for compatibility with parsing C headers. I'll probably leave it like this until a user emerges (and we can start using it) or someone complains (and we can drop it).